### PR TITLE
I believe this is ready for review.

### PR DESCRIPTION
**Pull request by jdlcdl**
I've tried to attack the 2 simplest tests, the ones you started and left incomplete for a noob.

But I'm not understanding what I'm doing wrong here.

### CODE REVIEW

**Comment by jdlcdl** [comment]
Exact error that I'm seeing at "# boom" comments is:

```
File "seedsigner-dev/src/seedsigner/gui/components.py", line 317, in __post_init__
    if not self.auto_line_break or full_text_width < self.supersampled_width - (2 * self.edge_padding * self.supersampling_factor):
TypeError: '<' not supported between instances of 'int' and 'MagicMock'
Appending next destination: UnhandledExceptionView({'error': ['TypeError', 'components.py, 317, in __post_init__', " '<' not supported between instances of 'int' and 'MagicMock'"]}) | clear_history: True
```

The line raising the exception has values where the expression becomes:
```
if not True or 75 < <MagicMock name='mock.Renderer.get_instance().canvas_width.__rmul__().__sub__()' id='140155864918640'>
```
...so I suspect that convincing this MagicMock instance to compare like an integer is the solution.

more later perhaps.

**Comment by jdlcdl** [comment]
As of #8984f23, tests pass on x86_64 Ubuntu as well as on my pi2.

My latest change is to merge dev into this pr because I knew it would create conflicts and I'm trying to get better about fixing those.  If you'd like Keith, I can `reset --hard` my local branch and `push --force` so that you can merge dev in when you're ready.

**Comment by kdmukai** [comment]
@jdlcdl I've also merged the latest from upstream/dev but I'm seeing some unexpected diffs that make me think we're somehow still slightly out of sync (see the diffs in docs/manual_installation.md).

I'm def not a git expert, but I think ideally your branch for this PR would rebase off my current "test_flows" branch (which now contains the latest from "upstream/dev") and only contain your proposed changes.

**Comment by kdmukai** [comment]
The diff makes it look like my test_flows branch is missing those latest changes from upstream/dev, but when I view the branch on its own outside of this PR, I see those changes applied as expected (e.g. the p2tr indentation fix in psbt_parser.py, line 151).

I have no clue why the diff in this PR seems to be reflecting an old state.

**Comment by jdlcdl** [comment]
Today, because my branch was out of sync with yours ('i'd previously merged 'dev' into mine), I did the following:
* local pr_339 branch: `git reset --hard 9c36f5c6608debc0a21a0a6e0f26291999ea7559` and
* local pr_339 branch: `git pull --rebase kdmukai test_flows`
* local pr_339 to my github repo: `git push --force`

All tests pass.

File: tests/test_flows_seed.py [hunk 1]
```diff
@@ -89,42 +91,181 @@ def test_invalid_mnemonic(self):
...
 
 
-    def test_export_xpub_flow(self):
+    def test_export_xpub_standard_flow(self):
         """
-            Selecting "Export XPUB" from the SeedOptionsView should enter the Export XPUB flow and end at the SeedOptionsView.
+            Selecting "Export XPUB" from the SeedOptionsView should enter the Export XPUB flow and end at the MainMenuView
         """
+
+        def test_standard_xpubs(sig_tuple, script_tuple, coord_tuple):
```
Comment (kdmukai) [inline tests/test_flows_seed.py]: Minor note: since the `test` prefix is a keyword for `pytest` to pickup tests, it's a tiny bit confusing/misleading to use the same naming convention in an internal function that `pytest` obviously will not try to run as an independent test.

In some cases I've used a leading underscore to specially identify internal functions, which is slightly unnecessary (no point in communicating "HANDS OFF!" if it's an internal function) but I also kind of like because my brain naturally just says, "Oh, this is an internal thing" when I see it invoked).

tldr: Just don't start the internal function name with "test" and we've settled at least 95% of the issue.

File: src/seedsigner/views/seed_views.py [hunk 2]
```diff
@@ -704,6 +707,9 @@ def run(self):
             selected_menu_num = 0
 
         else:
+            # imported locally instead of at module level to avoid flow-test warnings: "comparing int to MagicMock"
```
Comment (kdmukai) [inline src/seedsigner/views/seed_views.py]: This comment is too much detail for someone who's just interested in the main code. If the goal is to warn future flow-based test authors, adding a note in tests/README.md is probably a better place for it.

Maybe simplify to something like:
```
# imported locally to isolate hardware-dependent imports that cause problems in the test suite
```

File: tests/test_flows_seed.py [hunk 3]
```diff
@@ -89,42 +91,181 @@ def test_invalid_mnemonic(self):
...
+                    FlowStep(MainMenuView),
+                ]
+        )
+            
         # Load a finalized Seed into the Controller
         mnemonic = "blush twice taste dawn feed second opinion lazy thumb play neglect impact".split()
         self.controller.storage.set_pending_seed(Seed(mnemonic=mnemonic))
         self.controller.storage.finalize_pending_seed()
 
+        # these are (constant_value, display_name) tuples
```
Comment (kdmukai) [inline tests/test_flows_seed.py]: Suggested edit:
```python
# These are lists of tuples: (constant_value, display_name)
```

You could define their types, but it hurts readability so the above comment would still be helpful.
```python
sig_types: list[tuple[str, str]] = SettingsConstants.ALL_SIG_TYPES
```

File: src/seedsigner/views/seed_views.py (hunk 2 above)
Comment (jdlcdl) [reply src/seedsigner/views/seed_views.py]: I agree that this comment is best left in the test modules.  The same code pattern appears in your original work, in my work, and in Nick's work, so a comment in the test module, and just the code pattern is best in the actual code.

Ty!

File: tests/test_flows_seed.py (hunk 1 above)
Comment (jdlcdl) [reply tests/test_flows_seed.py]: Ty, yes, I see it now!

File: tests/test_flows_seed.py (hunk 3 above)
Comment (jdlcdl) [reply tests/test_flows_seed.py]: I'll prefer your suggestion because expressive code > comment, and I'll leave the comment too.

File: src/seedsigner/views/seed_views.py (hunk 2 above)
Comment (jdlcdl) [reply src/seedsigner/views/seed_views.py]: The test readme is succinct and to-the-point, while I'm wordy.  

For now, I've opted to extend the comment in tests/base.py where test-coders are most likely to check should they find this class not successfully mocked out.

I'll participate in the test/readme documentation changes once others have discussed what is desired for flow test docs.

File: tests/test_flows_seed.py [hunk 4]
```diff
@@ -89,42 +91,181 @@ def test_invalid_mnemonic(self):
...
+        self.settings.set_value(SettingsConstants.SETTING__SCRIPT_TYPES, [x for x,y in script_types if x!=disabled_script])
+        self.settings.set_value(SettingsConstants.SETTING__COORDINATORS, [x for x,y in coordinators if x!=disabled_coord])
+
+        # test that multisig is not an option via exception raised when NOT redirected to next step w/o a choice
+        with pytest.raises(FlowBasedTestUnexpectedViewError) as e:
+            self.run_sequence(
+                initial_destination_view_args=dict(seed_num=0),
+                sequence=[
+                    FlowStep(seed_views.SeedOptionsView, button_data_selection=seed_views.SeedOptionsView.EXPORT_XPUB),
+                    FlowStep(seed_views.SeedExportXpubSigTypeView, button_data_selection=disabled_sig),
```
Comment (kdmukai) [inline tests/test_flows_seed.py:175]: Oof, there's a subtle bug here. With multisig disabled, the only remaining sig type is single sig. Therefore that View should automatically redirect past this View and on to the next View (script type selection), which it does. But the test flow doesn't notice the expectation mismatch. The entry should be:
```
FlowStep(seed_views.SeedExportXpubSigTypeView, is_redirect=True)
```

I'll have to look at if I can do more rigorous validation on the button_data_selection; this sequence should fail if it's written to expect to be able to select a given button but the View redirects away without a user selection.

Comment (jdlcdl) [reply tests/test_flows_seed.py:175]: Darn.  testing is kinda hard...  it's not always that a test is a "good" test.  I was trying to prove that an exception would be raised, because indeed it should redirect when only one choice, which it does... but I was trying to fake the flow test into having a choice "disabled_sig" so it would fail and raise an Exception.  But it's not so expressive and clear of a test.

**Comment by kdmukai** [comment]
@jdlcdl I've updated the `test_flows` branch with a fix!

I also made some slight tweaks to your tests to work with the fixes. See my diff here:
https://github.com/kdmukai/seedsigner/commit/9759221c184770bddc669bb6c4757b8944c1c661

Can you review my latest changes and, if that all looks good, see if my changes to your tests make sense? If so, apply those changes to update this PR.

**Comment by jdlcdl** [comment]
Instead of holding you up, I've committed this morning's changes which use better precision of flow test exceptions... which also allowed deleting the last flow-step so that the exception is raised/caught exactly where a reader would expect it to be... (and not the previous step now that we've reached the wrong View -- which wasn't even the exact error that was being tested).

**Comment by kdmukai** [comment]
Really nice work here and exceptionally thorough! Thank you!
